### ISSUE TITLE
Another loss=NaN issue

### ISSUE THREAD

**Issue by cooperdk** [bmaltais/kohya_ss#724]
Picture says it all.

PLEASE CARRY OUT THE PROVIDED FIX SUGGESTIONS instead of just closing all issues without any response!

![image](https://user-images.githubusercontent.com/564150/236587613-5c743d8c-b3cf-48e0-8132-bc1ef6b84067.png)

**Comment by bmaltais** [comment]
I am not sure I understand. WHat should I fix? Can you create a Pull Request for the Fix? I think this is an issue that need to be resolved by kohya and not me. YOu do realise I only create the GUI and not the actual trainer code?

**Comment by cooperdk** [comment]
> I am not sure I understand. What should I fix? Can you create a Pull Request for the GUI portion that need to be fixed? I think this is an issue that need to be resolved by kohya and not me. You do realise I only create the GUI and not the actual trainer code?

I wasn't really thinking about that.
But you should not copy the entire tuner code into your own like you did. You should import it as a git dependency, making it possible to autoupdate when pulling your repository. That's best practice. :-)

**Comment by bmaltais** [comment]
I guess I don't follow best practices. Joking aside, there is a lot more history behind why things evolved to be what they are. It would indeed be possible to make the switch by reworking the code and repo structure but for now it is what it is

**Comment by cooperdk** [comment]
Its easy to just copy the code over and work out the requirements. And it turns out the actual training code is a mess that won't work on some of the most popular cards.
I can't make it work on my 3060 no matter what xformers or torch I install. I have years of Python knowledge.
I get black previews and nan losses, it might work, but the developer should have introduced a working fix to make the t-board work at least. He has all the tips in his issue list.
I am close to say screw it, Just gonna train in A1111 instead of figure something else out.

Second glance, your gui actually makes the script bearable. Too bad it doesn't work (or doesn't make sure your choices aren't invalid).

**Comment by bmaltais** [comment]
If you ever find what work for you 3060 ( I used to run on a 3060 and it was fine for kohya_ss for the most part) I can see what I can do to improve the installer to better support that card.

**Comment by bmaltais** [comment]
He uses the exact same versions as what is used by my installer... Not sure why it fail in your case... Maybe you have some global python module that conflict with the installer ... Where when you use his ore built venv you only get to use the local venv modules ... Quite possible

**Comment by cooperdk** [comment]
I think the issue on my end is due to the fact that I had actually installed venv using a Python 3.10.9 specific command, and expected the venv to "catch" a regular pip or python command. It doesn't. Instead, it ran all module installations and executed the code using Python 3.11 but with a Python 3.10 runtime in venv (which apparently isn't used).

Something is wrong with venv on Windows, it seems.

I am currently reinstalling the UI to see if the error is fixed by moving Py3.10.9 up in the path environment variables so it becomes the default.

**Comment by bmaltais** [comment]
@cooperdk Actually many SD AI projects out there for Windows only work on Python 3.10.6+ due to issues with 3.11

For example auto1111 only work with python 3.10.6+

**Comment by cooperdk** [comment]
No it doesn't. I have it working in Python 3.11 with torch 2.0 and my own compiled xformers-latest.

This script works fine for me now but the training previews look really bad. The data was previously used to train a hypernetwork in auto1111 and it worked fine then.

**Comment by rkfg** [comment]
This is really this GUI issue, unfortunately. I did some tests and tried to bisect to no avail, I though I knew the working commit but I was mistaken. Running on Debian Testing with 3090 Ti, training a LoRA:
![2023-05-28_16-54-01](https://github.com/bmaltais/kohya_ss/assets/184066/e8b453cf-8e35-4bc5-8517-88a4da9222c7)

The orange line is the original kohya scripts (current `main`), I use the same command the GUI prints. As you can see it trains normally for 30 epoch. All the other lines are from this GUI, at some point they skyrocket and become NaNs. The points are different because I didn't fix the seed but the issue happens every time at all commits I tried between `v21.5.5` and the current `master`. I also cleaned up `venv` and recreated with with `python3.10`, then ran `setup.sh`. It installed torch `1.12.1+cu116` and `0.0.14.dev0`, I ran training right after that, the issue still persists. Before that I manually upgraded torch to `2.0.1+cu118`, torchvision to `0.15.2+cu118` and xformers to `0.0.20.dev539+git.1d635e1` (unpacked the archive from anaconda because it's hard to get it from pip, they keep deleting old versions). The issue was still there.

For the `sd-scripts` repo I used torch `2.0.1`, torchvision `0.15.2` and same xformers `0.0.20.dev539+git.1d635e1` installed the same way. It trains normally to the 30th epoch. Perhaps there's something else that's different? The GUI worked before but I can't pinpoint the exact version.

Also, I suggest making the original `sd-script` a submodule so it's easier to update, if possible.

**Comment by cooperdk** [comment]
I made the submodule suggestion too, as it is logical and standardized.
Re the training difference, didn't test that, but it is interesting because the GUI just calls the kohya scripts. It means there must be an error in how the GUI calls the scripts (arguments).

**Comment by rkfg** [comment]
Not sure about the arguments, it prints the command before starting and I copypasted it to the directory with sd-scripts installed (and activated the venv there before that of course).

Experiment 1: I installed the torch versions with `+cu118` to `sd-scripts` explicitly, no idea if they're any different but for some reason the versions installed by `requirements.txt` don't have this prefix. Nevertheless, training goes just fine with the scripts.

Experiment 2: check if it's the GUI itself. Activate venv in `kohya_ss` and run the training command there without GUI. Result: the model collapses to nan.

Experiment 3: check if it's the dependencies. Keep the venv from `kohya_ss` activated but run the command in `sd-scripts`. Result: the model also collapses. So it must be something to do with the dependencies aside from torch and xformers. I suspect diffusers and accelerator.

**Comment by rkfg** [comment]
Okay, it was bitsandbytes. In https://github.com/bmaltais/kohya_ss/commit/7daa164d5357ad48a258e243f2c919cdd149d5d6#diff-4d7c51b1efe9043e44439a949dfd92e5827321b34082903477fd04876edb7552L8 you reverted it for windows but not for linux. I suggest setting it to 0.35.0 too, it fixes the issue with NaNs.
